### PR TITLE
[BEAM-9231] Annotate as Experimental/Internal missing classes in beam-sdks-java-core

### PR DESCRIPTION
**Pull request by iemejia**
Also make `SplittableDoFn` classes in `beam-sdks-java-core` Experimental and make Experimental annotations homogeneous.
For easier review take a look better at the separate commits.
R: @lukecwik

### CODE REVIEW

**Comment by iemejia** [comment]
Run Java PreCommit

**Comment by iemejia** [comment]
Run Java PreCommit

**Comment by iemejia** [comment]
Run Java PreCommit

**Comment by lukecwik** [comment]
I'm not sure if we need to mark runner libraries or anything outside of beam-sdks-java-core / beam-sdks-java-extensions-* / beam-sdks-java-io-* as `@Experimental`. WDYT?

**Comment by iemejia** [comment]
Agree 100%, I am going to remove the annotations from other places, e.g. `runners/* save for the `External` transform class that is misplaced there. For ref. [BEAM-8546](https://issues.apache.org/jira/browse/BEAM-8546)

**Comment by iemejia** [comment]
I added in the end 2 extra commits. First one just marks everything out of the modules we discussed out of Experimental/Internal and the second commit refines some `beam-sdks-java-core` classes Internal please take attention in detail to that last one in case you see any issue.

**Comment by iemejia** [comment]
Run JavaPortabilityApi PreCommit

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/External.java [hunk 1]
```diff
@@ -53,6 +55,7 @@
  * service. Note that this is a low-level API and mainly for internal use. A user may want to use
  * high-level wrapper classes rather than this one.
  */
+@Experimental(Kind.PORTABILITY)
```
Comment (lukecwik) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/External.java]: remove?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/harness/JvmInitializer.java [hunk 2]
```diff
@@ -33,6 +34,7 @@
  * automate this.
  */
 @Experimental
+@Internal
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/harness/JvmInitializer.java]: This isn't internal, we expect people to use this and for **workers** to honor it.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/DelegatingCounter.java [hunk 3]
```diff
@@ -17,12 +17,11 @@
...
 
-// Intentionally package private, this is meant to be used with Metrics.java and
-// LabeledMetrics.java
-
 import java.io.Serializable;
+import org.apache.beam.sdk.annotations.Internal;
 
 /** Implementation of {@link Counter} that delegates to the instance for the current context. */
+@Internal
 public class DelegatingCounter implements Metric, Counter, Serializable {
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/DelegatingCounter.java]: Can we make this package private like the comment said?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/CombineWithContext.java [hunk 4]
```diff
@@ -28,6 +29,7 @@
  *
  * <p>{@link CombineFnWithContext} is for users to extend.
  */
+@Internal
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/CombineWithContext.java]: I don't believe this is meant to be marked internal

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/ApiSurface.java [hunk 5]
```diff
@@ -84,6 +85,7 @@
  * not interesting.
  */
 @SuppressWarnings("rawtypes")
+@Internal
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/ApiSurface.java]: I believe the package doc states that all these util/* classes are internal but it might be wise to mark them `@Internal` anyways.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/testing/FakeBigQueryServices.java [hunk 6]
```diff
@@ -37,7 +38,7 @@
 import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.ImmutableList;
 
 /** A fake implementation of BigQuery's query service.. */
-@Experimental(Experimental.Kind.SOURCE_SINK)
+@Experimental(Kind.SOURCE_SINK)
```
Comment (lukecwik) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/testing/FakeBigQueryServices.java]: `@Internal` ?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/testing/FakeJobService.java [hunk 7]
```diff
@@ -87,7 +88,7 @@
 import org.joda.time.Duration;
 
 /** A fake implementation of BigQuery's job service. */
-@Experimental(Experimental.Kind.SOURCE_SINK)
+@Experimental(Kind.SOURCE_SINK)
```
Comment (lukecwik) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/testing/FakeJobService.java]: `@Internal` ?

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/External.java (hunk 1 above)
Comment (lukecwik) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/External.java]: remove since this is runners core?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/harness/JvmInitializer.java (hunk 2 above)
Comment (iemejia) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/harness/JvmInitializer.java]: For this PR I assumed that most things were `@Internal` if the audience was runner/harness authors instead of pipeline authors (final users). Is this class expected to be used by the latter? otherwise it still fits the `@Internal` definition like `DoFnInvokers` and the others, no?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/CombineWithContext.java (hunk 4 above)
Comment (iemejia) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/CombineWithContext.java]: ok removing it

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/ApiSurface.java (hunk 5 above)
Comment (iemejia) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/util/ApiSurface.java]: It does not but I am going to add that to the package. Probably worth to do that too for the schema package with the Experimental tag too.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/DelegatingCounter.java (hunk 3 above)
Comment (iemejia) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/DelegatingCounter.java]: Yes I tried to that I think and something failed, let me try again.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/testing/FakeBigQueryServices.java (hunk 6 above)
Comment (iemejia) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/testing/FakeBigQueryServices.java]: Definitely!

File: sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/DelegatingCounter.java (hunk 3 above)
Comment (iemejia) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/DelegatingCounter.java]: Oh I found `runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/LabeledMetrics.java` needs it so that's why it cannot be package private.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/testing/FakeJobService.java (hunk 7 above)
Comment (iemejia) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/testing/FakeJobService.java]: Yes

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/External.java (hunk 1 above)
Comment (iemejia) [reply runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/External.java]: I prefer to do an exception for this one because of https://issues.apache.org/jira/browse/BEAM-8546. This class is mislocated but nobody has cared about putting it on `sdks/java/core` where it belongs so this is to prevent than once it is moved there the 'experimental' annotation gets forgotten.

**Comment by iemejia** [comment]
Requested fixes done, I added an extra missing Experimental tag for `DoFn.OnTimersContext`). PTAL again @lukecwik

**Comment by iemejia** [comment]
Oups I have some more additions here (+ need to rebase this), will ping you back when ready again.

**Comment by iemejia** [comment]
Ok, added updates to tag Experimental/Internal at the package level and some other minor adjustments, again I suggest to look at the separate commits for an easier review. PTAL again @lukecwik (thanks).

**Comment by iemejia** [comment]
retest this please

**Comment by iemejia** [comment]
Run Seed Job

**Comment by iemejia** [comment]
Run Java PreCommit

**Comment by iemejia** [comment]
All green again. Can you PTAL @lukecwik

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/External.java (hunk 1 above)
Comment (lukecwik) [reply runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/External.java]: sgtm

File: sdks/java/core/src/main/java/org/apache/beam/sdk/harness/JvmInitializer.java (hunk 2 above)
Comment (lukecwik) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/harness/JvmInitializer.java]: We expect pipeline authors to write JVmInitializer and for the worker to use ServiceLoader to find them and use them.

**Comment by iemejia** [comment]
Thanks for the review Luke! I merged the PR manually to remove @Internal from `JvmInitializer`. One final question. What is the reason this class belongs in core instead of in harness?

**Comment by lukecwik** [comment]
`harness` is not meant to be depended on by users as its the main entry point for execution and pipeline authors have regularly requested the ability to be able to modify JVM parameters early on before their DoFn's or other parts of the system initialize.
